### ISSUE TITLE
Assessment logs should be written in Redis instead of the database

### ISSUE THREAD

**Issue by kelson42** [openzim/wp1#49]
That way we could:
* Speed up the writting/reading process of logs
* Secure easily that old logs are not saved for years without any reason
* Better SW architecture where temporary data are stored in temporary storage

**Comment by elfkuzco** [comment]
@audiodude , @kelson42 , I would like to work on this using Redis. Using it's native functionality to expire keys, we can also solve #778 by having a setting to define how long keys should stay.

Given the schema of the table does not reference any table as defined in the `schema.sql`, we can dump the data either as a json string with a key composed of the same values that define the primary key.

Another alternative would be to store it as a hash with the key and expire after interval.

Redis also supports storing JSON objects natively but JSON data grows much more faster than hash data type and I don't really see any advantage that it brings, especially for the log data where entries are not nested.

What do you think?

**Comment by audiodude** [comment]
Yes that sounds good, I think hashes are fine and we don't need JSON.

Please coordinate with @Shiv-aurora who has expressed interest in working on #778 (https://github.com/openzim/wp1/issues/778#issuecomment-2727191133).

**Comment by elfkuzco** [comment]
It turns out that there are a couple of places that involve querying for the logs using the project name and timestamp. Now, if we are to store logs on Redis as hashes, we are going to need to use RedisStack over native Redis as that comes with [search and query capabilities](https://redis.io/docs/latest/develop/interact/search-and-query/) and that would also mean migrating the data. Is this something you consider feasible @audiodude ? Or should we just fallback to the approach in #778 ?

**Comment by elfkuzco** [comment]
> Can you provide examples? As far as I can tell, we only query the logs when generating the log reports (besides a unit test which can be rewritten), and we do so for all logs for a given project. Can't we write the hash key as something like `'wp1-logs-Agriculture'`, where the project name is part of the key, and then generate the key on the fly to find all of the logs that way?

Yes, you are correct. The query was in a function that is only ever called in a unit test. I guess we can omit that and just fetch all the logs for a project. I generated the keys in the following way: "wp1-logs:{project}:{namespace}:{action}:{article}". This would allow querying in a way. It just doesn't offer the full capabilities of a SQL-like language. But I guess it works for this use case and should be perfectly fine.

**Comment by audiodude** [comment]
Sorry, could you better describe your intended schema for the Redis data? The insertion use case is that we want to log a particular action for an article in a project/namespace/article scope. We currently have:

```
        INSERT INTO logging
          (l_project, l_namespace, l_article, l_action, l_timestamp, l_old,
           l_new, l_revision_timestamp)
```

The retrieval use case is, get all actions for a given project/namespace. We sort the log entries by timestamp in application code, so that isn't necessary in the database

I don't think we want to be iterating and generating redis keys, right? We want to "grab" all the data for a project/namespace in one operation. Should we be using Redis lists? The JSON extension module?

Thanks!

**Comment by elfkuzco** [comment]
Okay. So, I want to store them as Redis hashes. Given we can call `attr.as_dict(log)`, that should be easy to setup.

Now, the key naming pattern is what would allow us to achieve a similar query that exists with SQL. If we use a key schema of `wp1-logs:{project}:{namespace}:{action}:{article}`, we can effectively make the following queries:

- query by project: `redis.scan(f"{wp1-logs:{project}:*}")`
- query by project and namespace: `redis.scan(f"{wp1-logs:{project}:{namespace}:*")`
- ...etc

The trick involves using the `*` to scan for all the keys following the pattern. We would be fetching all the hashes (logs) that match the pattern. Redis-py has a way to fetch the keys incrementally with a cursor and that should avoid the bottleneck of fetching all the keys at once.

My reason for suggesting Redis JSON or Redis Search was in the event we were filtering logs within a particular timestamp. The above approach does not allow filtering on a hash value but rather on a key namespace. That is the advantage Redis Search brings. But given we are not doing that, we can leverage the ability of tagging `*` along with the naming pattern. We just would be returning all the logs.

I hope this makes sense.

**Comment by audiodude** [comment]
This mostly makes sense. I wasn't aware of `redis.scan` with a wildcard parameter.

Where does the data live though? What gets set at those keys? We need:

```
  `l_timestamp` binary(14) NOT NULL,
  `l_old` varbinary(63) DEFAULT NULL,
  `l_new` varbinary(63) DEFAULT NULL,
  `l_revision_timestamp` binary(20) NOT NULL,
```

And I'm a bit worried that `action` is part of the key. Will we need to parse the key to get that data out? It seems hacky.

**Comment by elfkuzco** [comment]
Um, I am intending to store the whole data as part of the hash. This includes `l_project`, `l_namespace`, `l_action` in addition to those you mentioned.

This way, the action is stored in the hash too. If say we need to filter a particular action in a particular project e.g there's an action called `moved` in a project `A`, we can do:

`redis.scan(f"wp1-logs:{project}:*:moved:*")` 

I don't see any query like this though in the code though but by having all the fields in the hash and most (except timestamp values) as part of the key schema, we can get some flexibility with scanning. It's not perfect but it works.

**Comment by audiodude** [comment]
Okay so here is an example row from the production database:

```
MariaDB [enwp10_prod]> select * from logging limit 1;
+-----------+-------------+-----------+------------+----------------+------------+------------+----------------------+
| l_project | l_namespace | l_article | l_action   | l_timestamp    | l_old      | l_new      | l_revision_timestamp |
+-----------+-------------+-----------+------------+----------------+------------+------------+----------------------+
| .NET      |           0 | IKVM      | importance | 20250321000004 | NotA-Class | High-Class | 2025-03-20T07:19:02Z |
+-----------+-------------+-----------+------------+----------------+------------+------------+----------------------+
```

So are you using a hashmap? (`HMSET`) or something else?

Are you doing:

```
redis.set('wp1-logs:.Net:0:KVM:importance:20250321000004:NotA-Class:High-Class:2025-03-20T07:19:02Z', 1)
```

?

That seems really cumbersome to me.

**Comment by audiodude** [comment]
What about articles that have `:` in their name?

```
MariaDB [enwp10_prod]> SELECT * FROM logging WHERE l_project = 'Biography' AND l_article LIKE '%:%'
LIMIT 1;
+-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
| l_project | l_namespace | l_article             | l_action | l_timestamp    | l_old      | l_new          | l_revision_timestamp |
+-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
| Biography |           0 | 9/11:_The_Falling_Man | quality  | 20250103211304 | NotA-Class | Redirect-Class | 2025-01-03T09:20:44Z |
+-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
1 row in set (0.057 sec)
```

**Comment by elfkuzco** [comment]
> That seems really cumbersome to me.

The insertion is done in one place and doesn't need to be seen by callers. Typically, in my usage with Redis, I tend to define a class that defines methods for generating these names based on input parameters. This takes away the name generation from callers. But I understand your concerns.

BTW, we don't need to add the timestamp as part of the key name as I don't really see any benefit that it adds.

**Comment by elfkuzco** [comment]
> What about articles that have `:` in their name?
> 
> ```
> MariaDB [enwp10_prod]> SELECT * FROM logging WHERE l_project = 'Biography' AND l_article LIKE '%:%'
> LIMIT 1;
> +-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
> | l_project | l_namespace | l_article             | l_action | l_timestamp    | l_old      | l_new          | l_revision_timestamp |
> +-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
> | Biography |           0 | 9/11:_The_Falling_Man | quality  | 20250103211304 | NotA-Class | Redirect-Class | 2025-01-03T09:20:44Z |
> +-----------+-------------+-----------------------+----------+----------------+------------+----------------+----------------------+
> 1 row in set (0.057 sec)
> ```

Well, it's just a suggestion but I am open to any seperator you think would be fine. Different codebases have different requirements. Given Redis' flexibility, sticking with a convention/naming scheme helps makes things defined and stricter in a way.

**Comment by elfkuzco** [comment]
We don't need to use all the fields of the model for the key. Did a grep on the codebase and the only `WHERE` clause on the `logging` table is this:

`SELECT * FROM logging WHERE l_project = %s AND l_timestamp > %s`

This method is called in one place: `calculate_logs_to_update`. Now, it only queries for logs that have been generated within 7 days. But if we call `redis.expire(log_key, datetime.timedelta(days=7)` everytime we set a log at a particular key, we never have to worry about logs older than 7 days existing.

**Comment by audiodude** [comment]
Okay this all sounds pretty good. But I'm still curious, how would you implement a fictional method like:

```
def save_log(project, namespace, article, action, timestamp, old, new, rev_timestamp):
   pass
```

Just wondering what the actual data model is, because I'm not quite understanding the difference between storing the data in the key name versus what will actually be stored as the value for the key.

**Comment by elfkuzco** [comment]
> Okay this all sounds pretty good. But I'm still curious, how would you implement a fictional method like:
> 
> ```
> def save_log(project, namespace, article, action, timestamp, old, new, rev_timestamp):
>    pass
> ```
> 
> Just wondering what the actual data model is, because I'm not quite understanding the difference between storing the data in the key name versus what will actually be stored as the value for the key.


I'm going to save some typing and use the `Log` model as input to the function:
```py
def gen_redis_log(project, namespace, article, action) -> str:
  # this can be improved to do additional things such as using a defined prefix or seperator.
  # Moreover, a decorator can modify return value of this function by adding these constants
  # but for simplicity, I add them here
  return f"wp1:logs:{project}:{namespace}:{action}:{article}


def save_log(redis, log: Log):
  key = gen_redis_log_key(log.l_project, log.l_namespace, log.l_article, log.l_action)
  with redis.pipeline() as pipe:
    pipe.hset(key, mapping=attr.asdict(log))
    pipe.expire(key, datetime.timedelta(days=7), nx=True)
    pipe.execute()
```

All the fields of the `Log` exist on the hash. But because we have constrained the key to be strings which we can do exact matches on, we can do something like:
- get all logs for a project
- get all logs for a project with a given namespace
- get all logs for any project with a given namespace and article
- get all logs for any project with the article name

Because we have set up the generator function, we can supply `*` as the parameter name and that translates to a wildcard in the call to `redis.scan`

Getting a specific log would mean passing all the values to the generator function and using it in a call to `redis.scan`

Remember, the queries are always returning `all logs`. We still have some SQL-like filtering but it's not as fully-featured as native SQL in the sense that we can't do `OR` queries or `BETWEEN` to regulate timestamps. But given these are logs, they are fine tradeoffs. We never have logs more than 7 days, so no need for `BETWEEN` or `>`.

The only SQL-like filter we have is `AND` because of the way we set up the generator function. We are getting all the logs from the DB although `redis.scan` supports using a `count` amongst other params. But the order is never the same. We can use a cursor to fetch the logs incrementally (`redis.sscan`).

My reason for choosing those fields for the key names is so that we can do some sort of filtering on those exact values. Other fields like timestamp values are of no interest in the query since they are always returned from the hash.

**Comment by elfkuzco** [comment]
To demonstrate:
- get all logs for a project: `redis.scan(gen_redis_log_key(project, *, *, *)`
- get all logs for a project with namespace: `redis.scan(gen_redis_log_key(project, namespace, *, *)`
- get all logs for any project with a given namespace and article: `redis.scan(gen_redis_log_key(*, namespace, article, *)`
- get all logs for any project with the article name: `redis.scan(gen_redis_log_key(*, *, article, *)`

We can always add further parameters to `redis.scan` like type, count, etc

**Comment by audiodude** [comment]
Ok this sounds great! Thank you for the detailed explanation. Feel free to work on/mail a PR, the approach looks good.

**Comment by elfkuzco** [comment]
> Ok this sounds great! Thank you for the detailed explanation. Feel free to work on/mail a PR, the approach looks good.

I still need your input though on what would be a good seperator. Do you still have doubts about  `:`? Although, if you think about it, if we are searching for logs with a colon, it still works, because the function substitutes what it gets as input. The wildcard fits in perfectly where it should be weeding out articles that do not have that pattern.
